Export variants (master variant) during the first export

thisiscab opened the pull request:
The way that it works right now is that the exporter will export the
whole product, which means all of its variant.

Updaters will now only update self. So if we go in the admin and we
decide to update a unique product or a unique variant, then it will
update itself instead of saving the whole product will all of its
variant.

There is still some work to do in order to reduce the code that is
duplicated but that should still be good for CRing.

The goal is to run the exporter only once at the beginning of the
project, so fill the Shopify store with proper data. We were kind of
obligated to do that, since if we don't assign any variant to the
product when we create it, it assigns a dummy variant that is hard to
get rid of. I tried just to export the master variant when creating a
new product but it was really pain in the arse. This came out to be the
best solution I could come up with.

thisiscab committed:
First try on the master variant export

thisiscab committed:
Separate updaters and exporters

The way that it works right now is that the exporter will export the
whole product, which means all of its variant.

Updaters will now only update self. So if we go in the admin and we
decide to update a unique product or a unique variant, then it will
update itself instead of saving the whole product will all of its
variant.

There is still some work to do in order to reduce the code that is
duplicated but that should still be good for CRing.

The goal is to run the exporter only once at the beginning of the
project, so fill the Shopify store with proper data. We were kind of
obligated to do that, since if we don't assign any variant to the
product when we create it, it assigns a dummy variant that is hard to
get rid of. I tried just to export the master variant when creating a
new product but it was really pain in the arse. This came out to be the
best solution I could come up with.

braidn commented:
WIP, really should be used in a commit not as a commit. Can we quickly rebase, reword and force touch the first commit here @cabouffard ? Thanking 🙏

thisiscab committed:
Merge branch 'feature/shopify/variant-export' into feature/shopify/master-variant-export

thisiscab committed:
Remove legacy file

thisiscab committed:
Refactor the association saver

In multiple area of the exporter we have to save the id of the shopify
object to the associated spree object.

The AssociationSaver does it for us.

thisiscab committed:
Add forgotten attributes class to classes constructor

I have also decided to rename some variable, since I feel like this
would read better.

thisiscab commented:
I'd like to have a CR on this branch whenever is possible please!
Would anyone of you be able to do this? I'd like to get those branches out of the door ASAP :)

@braidn @DanielWright @bryanmtl

braidn commented on app/attributes/shopify/product_attributes.rb:
```diff
@@ -0,0 +1,41 @@
+    end
+
+    def attributes_for_variants
+      attributes = { variants: [] }
```
What would happen if we pushed this up to the initialize block?

braidn commented on app/converters/shopify/variant_converter.rb:
```diff
@@ -1,28 +1,38 @@
 module Shopify
   class VariantConverter
-    def initialize(variant:)
+    def initialize(variant:, aggregated: false)
```
The commit for this is across two different feature branches?

braidn commented on app/services/shopify/association_saver.rb:
```diff
@@ -0,0 +1,24 @@
+module Shopify
```
Cool! So in this branch this has been moved out to it's own file? I guess you can ignore the other comment in the other branch.

braidn commented on app/updaters/shopify/product_updater.rb:
```diff
@@ -0,0 +1,39 @@
+      shopify_product = find_shopify_product_for(spree_product)
+
+      if shopify_product.persisted?
+        shopify_product.update_attributes(product_attributes)
```
Can this be morphed into a private method and then the if else be set to a simple guard clause?

braidn commented on app/updaters/shopify/variant_updater.rb:
```diff
@@ -0,0 +1,29 @@
+      @attributor = attributor
+    end
+
+    def perform
```
💥 might have comment on this change for the same file in another branch and this may change it. Nice.

braidn commented on lib/tasks/shopify.rake:
```diff
@@ -0,0 +1,8 @@
+namespace :shopify do
+  task export_products: :environment do
```
Super simple entry point, love it.

bryanmtl commented:
@cabouffard This LGTM. I will play around with the code locally early next week to get a better feel for it. In the meantime, I don't see anything here that would prevent us from merging.

braidn reviewed (commented):
💥

thisiscab committed:
Change the perform action to make it smaller

As suggested in the CR, the logic is put in private methods so it's
easier to read the code and know what it does.
